File: core/src/main/java/net/twilightcity/flow/activity/MessageQueue.java
package net.twilightcity.flow.activity;

import net.twilightcity.flow.config.ModuleManager;
import net.twilightcity.gridtime.api.flow.activity.NewEditorActivityDto;
import net.twilightcity.gridtime.api.flow.activity.NewExecutionActivityDto;
import net.twilightcity.gridtime.api.flow.activity.NewExternalActivityDto;
import net.twilightcity.gridtime.api.flow.activity.NewModificationActivityDto;
import net.twilightcity.gridtime.api.flow.batch.NewFlowBatchEventDto;
import net.twilightcity.gridtime.api.flow.event.EventType;
import net.twilightcity.time.TimeService;

import java.io.File;
import java.io.FileWriter;
import java.io.IOException;
import java.io.PrintWriter;
import java.time.LocalDateTime;

public class MessageQueue {

    private ModuleManager moduleManager;
    private MessageLogger messageLogger;
    private TimeService timeService;

    public MessageQueue(TimeService timeService, ModuleManager moduleManager, File activeFlowFile) {
        this(new FileMessageLogger(activeFlowFile), timeService, moduleManager);
    }

    public MessageQueue(MessageLogger messageLogger, TimeService timeService, ModuleManager moduleManager) {
        this.messageLogger = messageLogger;
        this.timeService = timeService;
        this.moduleManager = moduleManager;
    }

    public void flush() {
        messageLogger.flush();
    }

    public void pushEditorActivity(Long durationInSeconds, String filePath, String module, boolean isModified) {
        pushEditorActivity(durationInSeconds, timeService.now(), filePath, module, isModified);
    }

    public void pushEditorActivity(Long durationInSeconds, LocalDateTime endTime, String filePath, String module, boolean isModified) {
        if (moduleManager.isModuleEnabled(module)) {
            NewEditorActivityDto activity = NewEditorActivityDto.builder()
                    .module(module)
                    .endTime(endTime)
                    .durationInSeconds(durationInSeconds)
                    .filePath(filePath)
                    .isModified(isModified)
                    .build();

            messageLogger.writeMessage(activity);
        }
    }

    public void pushModificationActivity(Long durationInSeconds, int modificationCount) {
        NewModificationActivityDto activity = NewModificationActivityDto.builder()
                .endTime(timeService.now())
                .durationInSeconds(durationInSeconds)
                .modificationCount(modificationCount)
                .build();

        messageLogger.writeMessage(activity);
    }

    public void pushExecutionActivity(Long durationInSeconds, String processName,
                                      int exitCode,
                                      String executionTaskType,
                                      boolean isDebug) {
        NewExecutionActivityDto activity = NewExecutionActivityDto.builder()
                .durationInSeconds(durationInSeconds)
                .endTime(timeService.now())
                .processName(processName)
                .exitCode(exitCode)
                .executionTaskType(executionTaskType)
                .isDebug(isDebug)
                .build();

        messageLogger.writeMessage(activity);
    }

    public void pushExternalActivity(Long durationInSeconds, String comment) {
        NewExternalActivityDto activity = NewExternalActivityDto.builder()
                .endTime(timeService.now())
                .durationInSeconds(durationInSeconds)
                .comment(comment)
                .build();

        messageLogger.writeMessage(activity);
    }

    public void pushEvent(EventType eventType, String message) {
        NewFlowBatchEventDto batchEvent = NewFlowBatchEventDto.builder()
                .position(timeService.now())
                .type(eventType)
                .comment(message)
                .build();

        messageLogger.writeMessage(batchEvent);
    }


    static class FileMessageLogger implements MessageLogger {
        private final File activeFlowFile;

        private final Object lock = new Object();
        private JSONConverter jsonConverter = new JSONConverter();

        FileMessageLogger(File activeFlowFile) {
            this.activeFlowFile = activeFlowFile;
        }

        public void flush() {
            //no-op since the FlowInsight app will be handling this now
        }

        public void writeMessage(Object message) {
            try {
                String messageAsJson = jsonConverter.toJSON(message);

                synchronized (lock) {
                    appendLineToFile(activeFlowFile, messageAsJson);
                }
            } catch (IOException ex) {
                throw new RuntimeException(ex);
            }
        }

        private void appendLineToFile(File file, String text) throws IOException {
            try (PrintWriter printWriter = new PrintWriter(new FileWriter(file, true))) {
                printWriter.println(text);
            }
        }



    }

}

File: vscode-plugin/src/activity/MessageQueue.ts
import { JSONConverter } from './JSONConverter';
import { TimeService } from '../time/TimeService';
import { ModuleManager } from '../config/ModuleManager';
import * as fs from 'fs';
import * as path from 'path';

export class MessageQueue {
    private jsonConverter = new JSONConverter();

    constructor(
        private timeService: TimeService,
        private moduleManager: ModuleManager,
        private activeFlowFile: string
    ) {
        // Ensure directory exists
        fs.mkdirSync(path.dirname(activeFlowFile), { recursive: true });
    }

    public pushEditorActivity(
        durationInSeconds: number,
        filePath: string,
        module: string | undefined,
        isModified: boolean
    ): void;
    public pushEditorActivity(
        durationInSeconds: number,
        endTime: Date,
        filePath: string,
        module: string | undefined,
        isModified: boolean
    ): void;
    public pushEditorActivity(
        durationInSeconds: number,
        endTimeOrFilePath: Date | string,
        filePathOrModule: string,
        moduleOrModified: string | undefined | boolean,
        isModified?: boolean
    ): void {
        const isEndTimeVersion = endTimeOrFilePath instanceof Date;
        
        const endTime = isEndTimeVersion ? endTimeOrFilePath : this.timeService.now();
        const filePath = isEndTimeVersion ? filePathOrModule : endTimeOrFilePath;
        const module = isEndTimeVersion ? moduleOrModified as string | undefined : filePathOrModule;
        const modified = isEndTimeVersion ? isModified : moduleOrModified as boolean;

        if (!module || this.moduleManager.isModuleEnabled(module)) {
            const activity = {
                durationInSeconds: Math.round(durationInSeconds),
                endTime,
                module,
                filePath,
                modified
            };
            this.writeMessage('EditorActivity', activity);
        }
    }

    public pushModificationActivity(durationInSeconds: number, modificationCount: number): void {
        const activity = {
            durationInSeconds: Math.round(durationInSeconds),
            endTime: this.timeService.now(),
            modificationCount
        };
        this.writeMessage('ModificationActivity', activity);
    }

    public pushExecutionActivity(
        durationInSeconds: number,
        processName: string,
        exitCode: number,
        executionTaskType: string,
        isDebug: boolean
    ): void {
        const activity = {
            durationInSeconds: Math.round(durationInSeconds),
            endTime: this.timeService.now(),
            processName,
            exitCode,
            executionTaskType,
            debug: isDebug
        };
        this.writeMessage('ExecutionActivity', activity);
    }

    public pushExternalActivity(durationInSeconds: number, comment: string): void {
        const activity = {
            durationInSeconds: Math.round(durationInSeconds),
            endTime: this.timeService.now(),
            comment
        };
        this.writeMessage('ExternalActivity', activity);
    }

    public pushEvent(eventType: string, message: string): void {
        const event = {
            type: eventType,
            position: this.timeService.now(),
            comment: message
        };
        this.writeMessage('Event', event);
    }

    public flush(): void {
        // No-op: The FlowPublisher will handle committing the active file
    }

    private writeMessage(_type: string, data: any): void {
        const message = `${this.jsonConverter.toJSON(data)}\n`;
        fs.appendFileSync(this.activeFlowFile, message);
    }
} 